GroupBy of stacked dim with strings renames underlying dims

Names for dimensions are lost (renamed) when they are stacked and grouped, if one of the dimensions has string coordinates.

```python
data = np.zeros((2,1,1))
dims = ['c', 'y', 'x']

d1 = xr.DataArray(data, dims=dims)
g1 = d1.stack(f=['c', 'x']).groupby('f').first()
print('Expected dim names:')
print(g1.coords)
print()

d2 = xr.DataArray(data, dims=dims, coords={'c': ['R', 'G']})
g2 = d2.stack(f=['c', 'x']).groupby('f').first()
print('Unexpected dim names:')
print(g2.coords)
```

#### Output
It is expected the 'f_level_0' and 'f_level_1' be 'c' and 'x', respectively in the second part below.
```
Expected dim names:
Coordinates:
  * f        (f) MultiIndex
  - c        (f) int64 0 1
  - x        (f) int64 0 0

Unexpected dim names:
Coordinates:
  * f          (f) MultiIndex
  - f_level_0  (f) object 'G' 'R'
  - f_level_1  (f) int64 0 0
```

#### Output of ``xr.show_versions()``
<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.7.4 (default, Jul  9 2019, 18:13:23) 
[Clang 10.0.1 (clang-1001.0.46.4)]
python-bits: 64
OS: Darwin
OS-release: 18.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8
libhdf5: 1.10.2
libnetcdf: 4.6.3

xarray: 0.12.3
pandas: 0.25.1
numpy: 1.17.1
scipy: 1.3.1
netCDF4: 1.5.2
pydap: None
h5netcdf: None
h5py: 2.9.0
Nio: None
zarr: None
cftime: 1.0.3.4
nc_time_axis: None
PseudoNetCDF: None
rasterio: None
cfgrib: None
iris: None
bottleneck: None
dask: None
distributed: None
matplotlib: 3.1.1
cartopy: None
seaborn: None
numbagg: None
setuptools: 41.2.0
pip: 19.2.3
conda: None
pytest: None
IPython: 7.8.0
sphinx: None
</details>

Thanks @max-sixty.  Contrary to my warning about not doing a PR, I couldn't help myself and dug in a bit.  It turns out that string coordinates aren't the problem, it's when the coordinate isn't in sorted order.  For example, @chrisroat's original example doesn't error if the coordinate is `["G", "R"]` instead of `["R", "G"]`.  A more concrete WIP test:

```python
def test_stack_groupby_unsorted_coord():
    data = [[0, 1], [2, 3]]
    data_flat = [0, 1, 2, 3]
    dims = ["y", "x"]
    y_vals = [2, 3]

    # "y" coord is in sorted order, and everything works
    arr = xr.DataArray(data, dims=dims, coords={"y": y_vals})
    actual1 = arr.stack(z=["y", "x"]).groupby("z").first()
    midx = pd.MultiIndex.from_product([[2, 3], [0, 1]], names=dims)
    expected1 = xr.DataArray(data_flat, dims=["z"], coords={"z": midx})
    xr.testing.assert_equal(actual1, expected1)
    
    # Now "y" coord is NOT in sorted order, and the bug appears
    arr = xr.DataArray(data, dims=dims, coords={"y": y_vals[::-1]})
    actual2 = arr.stack(z=["y", "x"]).groupby("z").first()
    midx = pd.MultiIndex.from_product([[3, 2], [0, 1]], names=dims)
    expected2 = xr.DataArray(data_flat, dims=["z"], coords={"z": midx})
    xr.testing.assert_equal(actual2, expected2)

test_stack_groupby_str_coords()
```
yields
```python
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)

[...]

AssertionError: Left and right DataArray objects are not equal

Differing values:
L
    array([2, 3, 0, 1])
R
    array([0, 1, 2, 3])
Differing coordinates:
L * z        (z) MultiIndex
  - z_leve...(z) int64 2 2 3 3
  - z_leve...(z) int64 0 1 0 1
R * z        (z) MultiIndex
  - y        (z) int64 3 3 2 2
  - x        (z) int64 0 1 0 1
```

I'll return to this tomorrow, in the meantime if this triggers any thoughts about the best path forward, that would be much appreciated!

--- a/doc/whats-new.rst
+++ b/doc/whats-new.rst
@@ -39,7 +39,9 @@ New Features
 
 Bug fixes
 ~~~~~~~~~
-
+- Fix renaming of coords when one or more stacked coords is not in
+  sorted order during stack+groupby+apply operations. (:issue:`3287`,
+  :pull:`3906`) By `Spencer Hill <https://github.com/spencerahill>`_
 
 Documentation
 ~~~~~~~~~~~~~
@@ -83,13 +85,13 @@ New Features
 - Added support for :py:class:`pandas.DatetimeIndex`-style rounding of
   ``cftime.datetime`` objects directly via a :py:class:`CFTimeIndex` or via the
   :py:class:`~core.accessor_dt.DatetimeAccessor`.
-  By `Spencer Clark <https://github.com/spencerkclark>`_ 
+  By `Spencer Clark <https://github.com/spencerkclark>`_
 - Support new h5netcdf backend keyword `phony_dims` (available from h5netcdf
   v0.8.0 for :py:class:`~xarray.backends.H5NetCDFStore`.
   By `Kai Mühlbauer <https://github.com/kmuehlbauer>`_.
 - Add partial support for unit aware arrays with pint. (:pull:`3706`, :pull:`3611`)
   By `Justus Magin <https://github.com/keewis>`_.
-- :py:meth:`Dataset.groupby` and :py:meth:`DataArray.groupby` now raise a 
+- :py:meth:`Dataset.groupby` and :py:meth:`DataArray.groupby` now raise a
   `TypeError` on multiple string arguments. Receiving multiple string arguments
   often means a user is attempting to pass multiple dimensions as separate
   arguments and should instead pass a single list of dimensions.
@@ -107,7 +109,7 @@ New Features
   By `Maximilian Roos <https://github.com/max-sixty>`_.
 - ``skipna`` is available in :py:meth:`Dataset.quantile`, :py:meth:`DataArray.quantile`,
   :py:meth:`core.groupby.DatasetGroupBy.quantile`, :py:meth:`core.groupby.DataArrayGroupBy.quantile`
-  (:issue:`3843`, :pull:`3844`) 
+  (:issue:`3843`, :pull:`3844`)
   By `Aaron Spring <https://github.com/aaronspring>`_.
 
 Bug fixes
@@ -801,7 +803,7 @@ Bug fixes
 Documentation
 ~~~~~~~~~~~~~
 
-- Created a `PR checklist <https://xarray.pydata.org/en/stable/contributing.html/contributing.html#pr-checklist>`_ 
+- Created a `PR checklist <https://xarray.pydata.org/en/stable/contributing.html/contributing.html#pr-checklist>`_
   as a quick reference for tasks before creating a new PR
   or pushing new commits.
   By `Gregory Gundersen <https://github.com/gwgundersen>`_.

--- a/xarray/core/groupby.py
+++ b/xarray/core/groupby.py
@@ -371,9 +371,7 @@ class GroupBy(SupportsArithmetic):
 
             # look through group to find the unique values
             group_as_index = safe_cast_to_index(group)
-            sort = bins is None and (
-                not isinstance(group_as_index, pd.MultiIndex)
-            )
+            sort = bins is None and (not isinstance(group_as_index, pd.MultiIndex))
             unique_values, group_indices = unique_value_groups(
                 group_as_index, sort=sort
             )

--- a/xarray/core/groupby.py
+++ b/xarray/core/groupby.py
@@ -370,10 +370,11 @@ class GroupBy(SupportsArithmetic):
                 group = group.dropna(group_dim)
 
             # look through group to find the unique values
-            all_sorted = all([safe_cast_to_index(c).is_monotonic_increasing
-                              for c in group.coords.values()])
+            sort = bins is None and (
+                not isinstance(safe_cast_to_index(group), pd.MultiIndex)
+            )
             unique_values, group_indices = unique_value_groups(
-                safe_cast_to_index(group), sort=(bins is None and all_sorted)
+                safe_cast_to_index(group), sort=sort
             )
             unique_coord = IndexVariable(group.name, unique_values)
 

--- a/xarray/core/groupby.py
+++ b/xarray/core/groupby.py
@@ -370,11 +370,12 @@ class GroupBy(SupportsArithmetic):
                 group = group.dropna(group_dim)
 
             # look through group to find the unique values
+            group_as_index = safe_cast_to_index(group)
             sort = bins is None and (
-                not isinstance(safe_cast_to_index(group), pd.MultiIndex)
+                not isinstance(group_as_index, pd.MultiIndex)
             )
             unique_values, group_indices = unique_value_groups(
-                safe_cast_to_index(group), sort=sort
+                group_as_index, sort=sort
             )
             unique_coord = IndexVariable(group.name, unique_values)
 

--- a/xarray/core/groupby.py
+++ b/xarray/core/groupby.py
@@ -370,8 +370,10 @@ class GroupBy(SupportsArithmetic):
                 group = group.dropna(group_dim)
 
             # look through group to find the unique values
+            all_sorted = all([safe_cast_to_index(c).is_monotonic_increasing
+                              for c in group.coords.values()])
             unique_values, group_indices = unique_value_groups(
-                safe_cast_to_index(group), sort=(bins is None)
+                safe_cast_to_index(group), sort=(bins is None and all_sorted)
             )
             unique_coord = IndexVariable(group.name, unique_values)
 

--- a/xarray/tests/test_dataarray.py
+++ b/xarray/tests/test_dataarray.py
@@ -1208,14 +1208,13 @@ class TestDataArray:
         dims = ["x", "y"]
         y_vals = [2, 3]
 
-        # "y" coord is in sorted order, and everything works
         arr = xr.DataArray(data, dims=dims, coords={"y": y_vals})
         actual1 = arr.stack(z=dims).groupby("z").first()
         midx1 = pd.MultiIndex.from_product([[0, 1], [2, 3]], names=dims)
         expected1 = xr.DataArray(data_flat, dims=["z"], coords={"z": midx1})
         xr.testing.assert_equal(actual1, expected1)
 
-        # Now "y" coord is NOT in sorted order, and the bug appears
+        # GH: 3287.  Note that y coord values are not in sorted order.
         arr = xr.DataArray(data, dims=dims, coords={"y": y_vals[::-1]})
         actual2 = arr.stack(z=dims).groupby("z").first()
         midx2 = pd.MultiIndex.from_product([[0, 1], [3, 2]], names=dims)

--- a/xarray/tests/test_dataarray.py
+++ b/xarray/tests/test_dataarray.py
@@ -1202,6 +1202,26 @@ class TestDataArray:
         expected = data.isel(xy=[0, 1]).unstack("xy").squeeze("y").drop_vars("y")
         assert_equal(actual, expected)
 
+    def test_stack_groupby_unsorted_coord(self):
+        data = [[0, 1], [2, 3]]
+        data_flat = [0, 1, 2, 3]
+        dims = ["x", "y"]
+        y_vals = [2, 3]
+
+        # "y" coord is in sorted order, and everything works
+        arr = xr.DataArray(data, dims=dims, coords={"y": y_vals})
+        actual1 = arr.stack(z=dims).groupby("z").first()
+        midx1 = pd.MultiIndex.from_product([[0, 1], [2, 3]], names=dims)
+        expected1 = xr.DataArray(data_flat, dims=["z"], coords={"z": midx1})
+        xr.testing.assert_equal(actual1, expected1)
+
+        # Now "y" coord is NOT in sorted order, and the bug appears
+        arr = xr.DataArray(data, dims=dims, coords={"y": y_vals[::-1]})
+        actual2 = arr.stack(z=dims).groupby("z").first()
+        midx2 = pd.MultiIndex.from_product([[0, 1], [3, 2]], names=dims)
+        expected2 = xr.DataArray(data_flat, dims=["z"], coords={"z": midx2})
+        xr.testing.assert_equal(actual2, expected2)
+
     def test_virtual_default_coords(self):
         array = DataArray(np.zeros((5,)), dims="x")
         expected = DataArray(range(5), dims="x", name="x")

